nomad/plan_apply.go

```diff
@@ -231,8 +240,41 @@ func evaluatePlan(snap *state.StateSnapshot, plan *structs.Plan) (*structs.PlanR
 		if nodeAlloc := plan.NodeAllocation[nodeID]; len(nodeAlloc) > 0 {
 			result.NodeAllocation[nodeID] = nodeAlloc
 		}
+		return
+	}
+
+	// Get the pool channels
+	req := pool.RequestCh()
+	resp := pool.ResultCh()
+	outstanding := 0
+	didCancel := false
+
+	// Evalute each node in the plan, handling results as
+	// they are ready to avoid blocking.
+	for nodeID := range nodeIDs {
+		select {
+		case req <- evaluateRequest{snap, plan, nodeID}:
+			outstanding++
+		case r := <-resp:
+			outstanding--
+			if cancel := handleResult(r.nodeID, r.fit, r.err); cancel {
+				didCancel = true
+				break
+			}
+		}
+	}
+
+	// Drain the remaining results
+	for outstanding > 0 {
+		r := <-resp
+		if !didCancel {
```

Comment:
I would just put a comment that if one has cancelled we can save ourselves work by not handling the results